fix: role_arn support in AWSCloudWatch and AWSEmf exporters

yaten2302 opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description

This PR attempts to fix re-add the support of `role_arn` in AWSCloudWatch and AWSEmf exporters.
The support was present in aws-go-sdk-v1, but broke while moving to v2.

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Fixes #42115

yaten2302 committed:
fix: role_arn support in AWSCloudWatch and AWSEmf exporters

Signed-off-by: Yaten <yaten598@gmail.com>

yaten2302 committed:
add changelog entry & tidy go.mod

Signed-off-by: Yaten <yaten598@gmail.com>

axw commented on internal/aws/awsutil/conn.go:
```diff
@@ -103,6 +105,21 @@ func GetAWSConfig(ctx context.Context, logger *zap.Logger, settings *AWSSessionS
 		return aws.Config{}, err
 	}
 
+	if settings.RoleARN != "" {
```
Is it possible to add a test for this? e.g. by mocking both the STS and the CloudWatch API

It was broken once before, adding a test will help avoid breaking it once again

pavolloffay replied:
They suggest using mocking https://docs.aws.amazon.com/sdk-for-go/v2/developer-guide/unit-testing.html 

but I am not sure this can validate the auth mechanism.

axw reviewed (commented):
I'm afraid I don't have enough expertise in the AWS SDK to know if this is right. A few things that may help:
 - documenting how you have tested this
 - adding unit tests
 - linking the old code related to assume role, that was working with aws-sdk-go

yaten2302 commented:
@axw , addressing your comments:

1. I haven't tested this component, because, while I was going through the codebase for v1 and comparing the aws go sdk v1 and v2 changes in the docs, I found that there were some internal changes made as a result of which we would now require to mention about the `RoleARN` explicitly (in v2).
2. I'll add the unit tests. This would require refactoring of the [`GetAWSConfig`](https://github.com/yaten2302/opentelemetry-collector-contrib/blob/fix/support-for-role_arn-in-cloud-watch/internal/aws/awsutil/conn.go#L88).
3. Previously (till version 0.123), this `GetAWSConfig` func was named as [`GetAWSConfigSession`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/release/v0.123.x/internal/aws/awsutil/conn.go#L110-L161). And if we check the [`148th line`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/release/v0.123.x/internal/aws/awsutil/conn.go#L148) of this func, we've `cn.newAWSSession(logger, cfg.RoleARN, cfg.ExternalID, awsRegion)`. In this, we're passing the `cfg.RoleARN`. But, this support was removed in v2. So, now we need to explicitly mention that in our func. This has also been implemented in [`AWSKinesisExporter (aws-go-sdk-v2)`](https://github.com/yaten2302/opentelemetry-collector-contrib/blob/fix/support-for-role_arn-in-cloud-watch/exporter/awskinesisexporter/exporter.go). If we check from lines - [65-73](https://github.com/yaten2302/opentelemetry-collector-contrib/blob/fix/support-for-role_arn-in-cloud-watch/exporter/awskinesisexporter/exporter.go#L65-L73).

axw commented:
> This has also been implemented in [AWSKinesisExporter (aws-go-sdk-v2)](https://github.com/yaten2302/opentelemetry-collector-contrib/blob/fix/support-for-role_arn-in-cloud-watch/exporter/awskinesisexporter/exporter.go). If we check from lines - [65-73](https://github.com/yaten2302/opentelemetry-collector-contrib/blob/fix/support-for-role_arn-in-cloud-watch/exporter/awskinesisexporter/exporter.go#L65-L73).

Thanks @yaten2302. Based on that, assuming it works for Kinesis, I suppose the code change is fine. If nothing else, a little bit of manual testing would be good.

yaten2302 commented:
@axw , I was trying to refactor the `GetAWSConfig()` func, so for this I was checking the docs and all.
So, should I go ahead with the adding the unit tests or manual testing would serve the purpose?

yaten2302 commented:
@axw , if I've understood this correctly, then, manual testing is a must for this PR, so that it doesn't breaks in future, right?
I also haven't used this component or AWSCloudWatch. So, maybe we can ask those folks who actually depend this component.
But, is manual testing (on real logs/data on AWS cloud-watch or emf), really possible locally?

WDYT?

axw commented:
> if I've understood this correctly, then, manual testing is a must for this PR, so that it doesn't breaks in future, right?

Manual testing is just to make sure the change does what you intended. That code can unintentionally removed or broken, so it doesn't prevent future breakage.

Automated tests will help prevent future breakage. Ideally you should have some [functional tests](https://en.wikipedia.org/wiki/Functional_testing) that would fail if the `role_arn` support breaks.

> I also haven't used this component or AWSCloudWatch. So, maybe we can ask those folks who actually depend this component.

Maybe @pavolloffay?

> But, is manual testing (on real logs/data on AWS cloud-watch or emf), really possible locally?

I don't see why not. If it's not possible, or not practical, you could also run the collector in an EC2 instance.

pavolloffay commented on .chloggen/fix_support-for-role_arn-in-cloud-watch.yaml:
```diff
@@ -0,0 +1,27 @@
+change_type: "breaking"
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: "awscloudwatchlogexporter"
```
please list here all affected components: these two are missing: `awsemfexporter`, `awsxrayexporter`

yaten2302 replied:
Done!!

pavolloffay commented on .chloggen/fix_support-for-role_arn-in-cloud-watch.yaml:
```diff
@@ -0,0 +1,27 @@
+component: "awscloudwatchlogexporter"
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: "re-add support for `role_arn` in `awscloudwatchlogexporter` and `awsemfexporter`."
```
Fix support for `role_arn` (STS, short-lived token authentication).

yaten2302 replied:
Updated this!

pavolloffay commented:
If anyone wants to test this I have built a docker image `pavolloffay/otelcol-contrib:fix-aws-rolearn2`

andrzej-stencel reviewed (commented):
Thanks @yaten2302 for the contribution, much appreciated. Also thank you everyone involved in testing this manually.

We should have a unit test for this if possible.

Other than that, can we get the eyes of some AWS subject matter experts on this before merging? @mxiamxia @Aneurysm9 can you take a look as codeowners?

yaten2302 committed:
Merge branch 'main' into fix/support-for-role_arn-in-cloud-watch

yaten2302 committed:
update changelog entry

Signed-off-by: Yaten <yaten598@gmail.com>

yaten2302 committed:
update testify pkg version

Signed-off-by: Yaten <yaten598@gmail.com>

yaten2302 committed:
Merge branch 'main' into fix/support-for-role_arn-in-cloud-watch

yaten2302 committed:
refactor GetAWSConfig + update changelog entry

Signed-off-by: Yaten <yaten598@gmail.com>

yaten2302 committed:
tidy go.mod files

Signed-off-by: Yaten <yaten598@gmail.com>

yaten2302 commented:
@andrzej-stencel @pavolloffay , wanted to ask here that, after [this](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/42541/commits/52b0d2238713d0d9925be5058c2be5c00581a473) commit, there're a lot of failing tests because, I added a new param in the `GetAWSConfig()` func. So, should I go ahead with the refactor or should I revert back to previous commit?
This refactor was made as an attempt to add the unit tests. Though, we've checked the docker img of the collector and which is working fine, just wanted your reviews on this that should I go ahead with writing the tests and the further refactor or revert to prev commits in which there're no new params?

axw commented on internal/aws/awsutil/conn.go:
```diff
@@ -83,7 +89,7 @@ func getProxyURL(finalProxyAddress string) (*url.URL, error) {
 
 // GetAWSConfig returns AWS config instance.
-func GetAWSConfig(ctx context.Context, logger *zap.Logger, settings *AWSSessionSettings) (aws.Config, error) {
+func GetAWSConfig(ctx context.Context, logger *zap.Logger, settings *AWSSessionSettings, stsClient STSClient) (aws.Config, error) {
```
It looks like you added the stsClient parameter so the client can be mocked, but you're only mocking it in this package.

I suggest you revert this one, and introduce a new unexported method for testing the STS part within this package. That would keep the API clean for the exporters.

axw commented on .chloggen/fix_support-for-role_arn-in-cloud-watch.yaml:
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: "breaking"
```
Why is this a breaking change?

yaten2302 replied:
The `role_arn` support was breaking👀?

yaten2302 replied:
Or maybe this can be put under `bug fix`?

axw replied:
Yes, `bug_fix` is appropriate here. `breaking` means that your change breaks something.

yaten2302 replied:
done 👍

yaten2302 committed:
Merge branch 'main' into fix/support-for-role_arn-in-cloud-watch

yaten2302 committed:
remove stsClient param from GetAWSConfig + update changelog entry

Signed-off-by: Yaten <yaten598@gmail.com>

yaten2302 committed:
update awsxray component acc to GetAWSConfig

Signed-off-by: Yaten <yaten598@gmail.com>

yaten2302 committed:
update role_arn tests

Signed-off-by: Yaten <yaten598@gmail.com>

yaten2302 committed:
tidy go.mod files

Signed-off-by: Yaten <yaten598@gmail.com>

yaten2302 committed:
attempt to fix linter errors

Signed-off-by: Yaten <yaten598@gmail.com>

yaten2302 committed:
Merge branch 'main' into fix/support-for-role_arn-in-cloud-watch

yaten2302 committed:
fix linter issues

Signed-off-by: Yaten <yaten598@gmail.com>

yaten2302 commented:
In CI, it's failing. But on locally it's working fine.
Is there some other behaviour on your side?

yaten2302 committed:
Merge branch 'main' into fix/support-for-role_arn-in-cloud-watch

atoulme committed:
Merge branch 'main' into fix/support-for-role_arn-in-cloud-watch

yaten2302 committed:
Merge branch 'main' into fix/support-for-role_arn-in-cloud-watch

yaten2302 committed:
refactor GetAWSConfig to add a helper func for better unit testing

Signed-off-by: Yaten <yaten598@gmail.com>

yaten2302 committed:
Merge branch 'main' into fix/support-for-role_arn-in-cloud-watch

yaten2302 committed:
attempt to fix CI checks

Signed-off-by: Yaten <yaten598@gmail.com>

yaten2302 committed:
Merge branch 'main' into fix/support-for-role_arn-in-cloud-watch

yaten2302 committed:
Merge branch 'main' into fix/support-for-role_arn-in-cloud-watch

yaten2302 committed:
Merge branch 'main' into fix/support-for-role_arn-in-cloud-watch
